databuilder/query_model/transforms.py

```diff
@@ -41,39 +97,27 @@ class PickOneRowPerPatientWithColumns(PickOneRowPerPatient):
 
 
 def apply_transforms(variables):
-    # For algorithmic ease we're going to be mutating the query objects. Since the QM
-    # graph is a notionally immutable structure consisting of "frozen" dataclasses, we need
-    # to do that carefully. In particular:
-    #
-    # 1. We copy the data that is passed in so that callers don't observe side-effects.
-    # 2. We copy the results on the way out to rectify the state of containers that depend
-    #    on object hashes which may have changed unexpectedly (specifically the frozenset used
-    #    to hold selected columns).
-    # 3. When we store QM nodes in equality/hash-sensitive containers during this manipulation
-    #    we use customized versions of those containers which ignore the __eq__() and
-    #    __hash__() implementations provided by dataclasses.
-    variables = copy.deepcopy(variables)
     nodes = all_nodes_from_variables(variables)
     reverse_index = build_reverse_index(nodes)
 
     transforms = [
         (PickOneRowPerPatient, rewrite_sorts),
     ]
 
+    rewriter = QueryGraphRewriter()
     for type_, transform in transforms:
-        apply_transform(type_, transform, nodes, reverse_index)
+        apply_transform(rewriter, type_, transform, nodes, reverse_index)
 
-    variables = copy.deepcopy(variables)  # see comment above
-    return variables
+    return rewriter.rewrite(variables)
```

Comment:
This won't be correct for non-commutative transformations. Obviously that's fine right now when we only have one transformation, but perhaps it deserves a comment to this effect on the `transforms` list.

Reply:
Ah yes, my thought was that we might just want to do one rewrite per transform. They're relatively cheap now (and can short-circuit if there are no replacements) and that's probably easier than worrying about which of our transforms commute with which. Definitely worth a comment though.

Reply:
I've rebased a comment into the first commit.